skimage/util/_remap.pyx

```diff
@@ -9,14 +12,13 @@ def _map_array(np_anyint[:] inarr, np_numeric[:] outarr,
     # build the map from the input and output vectors
     cdef size_t i, n_map, n_array
     cdef unordered_map[np_anyint, np_numeric] lut
+    cdef unordered_map[np_anyint, np_numeric].iterator it
     n_map = inval.shape[0]
     for i in range(n_map):
         lut[inval[i]] = outval[i]
     # apply the map to the array
     n_array = inarr.shape[0]
-    # The prange option gave some compilation warnings
-    #  "Unsigned index type not allowed before OpenMP 3.0"
-    # and didn't seem to be any faster
-    # for i in prange(n_array, nogil=True): #
-    for i in range(n_array):
-        outarr[i] = lut[inarr[i]]
+    for i in prange(n_array, nogil=True): #
+        it = lut.find(inarr[i])
+        if it != lut.end():
+            outarr[i] = dereference(it).second
```

Comment:
Re

Technically, the new handling for unknown values in `input_arr` introduces a breaking API change in case `out` is not `None`. E.g. previously (on `main`)
```python
import numpy as np
import skimage as ski
ski.util.map_array(
    np.array([1, 1, 2, 2]),
    np.array([1]),
    np.array([3]),
    np.array([4, 4, 4, 4])
)
```
returns `array([3, 3, 0, 0])`. This PR currently gives `array([3, 3, 4, 4])` instead.

I propose to just add an else clause like this and revert the initialization above if `out=None` back to `np.empty`.

```suggestion
            outarr[i] = dereference(it).second
        else:
            outarr[i] = 0
```

Reply:
If this has performance implications we can revert it once we have a mechanism like `default_out_value=0` (see https://github.com/scikit-image/scikit-image/pull/7266#issuecomment-1854863917) in place.

Reply:
My intuition is that this will be slower. But this may be good for consistency.

Reply:
Very good catch @lagru! I agree that this is a significant change in behaviour so we should avoid it.

Reply:
(And we should add a test for it.)